File: artifacts/libzahl-1.0/translate/src/ztrunc.rs
use crate::*;

pub fn ztrunc(a: &mut Z, b: &Z, bits: usize) {
    let mut mask: ZahlCharT = 1;
    let chars: usize;
    let mut i: usize;

    if zzero(b) {
        a.sign = 0;
        return;
    }

    chars = (bits + (32 - 1)) >> 5;
    a.sign = b.sign;
    a.used = if chars < b.used { chars } else { b.used };
    
    let mut bits = if a.used < chars { 0 } else { bits };
    
    if a as *const _ != b as *const _ {
        if a.alloced < a.used {
            libzahl_realloc(a, a.used);
        }
        
        if let Some(a_chars) = &mut a.chars {
            if let Some(b_chars) = &b.chars {
                a_chars[..a.used].copy_from_slice(&b_chars[..a.used]);
            }
        }
    }

    bits = bits & (32 - 1);
    if bits != 0 {
        mask <<= bits;
        mask -= 1;
        if let Some(a_chars) = &mut a.chars {
            if a.used > 0 {
                a_chars[a.used - 1] &= mask;
            }
        }
    }

    i = a.used;
    while i > 0 {
        i -= 1;
        if let Some(a_chars) = &a.chars {
            if a_chars[i] != 0 {
                return;
            }
        }
    }

    a.sign = 0;
}

File: artifacts/libzahl-1.0/decompose/src/ztrunc.c
void ztrunc(z_t a, z_t b, size_t bits)
{
  zahl_char_t mask = 1;
  size_t chars;
  size_t i;
  if (zzero(b))
  {
    a->sign = 0;
    return;
  }
  chars = (bits + (32 - 1)) >> 5;
  a->sign = b->sign;
  a->used = (chars < b->used) ? (chars) : (b->used);
  if (a->used < chars)
  {
    bits = 0;
  }
  if (a != b)
  {
    do
    {
      if (a->alloced < a->used)
      {
        libzahl_realloc(a, a->used);
      }
    }
    while (0);
    memcpy(a->chars, b->chars, a->used * (sizeof(zahl_char_t)));
  }
  bits = bits & (32 - 1);
  if (bits)
  {
    mask <<= bits;
    mask -= 1;
    a->chars[a->used - 1] &= mask;
  }
  for (i = a->used; i--;)
    if (a->chars[i])
  {
    return;
  }

  a->sign = 0;
}


/*** DEPENDENCIES:
inline static int zzero(z_t a)
{
  return !a->sign;
}


----------------------------
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1]
----------------------------
void libzahl_realloc(z_t a, size_t need)
{
  size_t i;
  size_t x;
  zahl_char_t *new;
  unsigned int new_idx = 0;
  if ((need & ((~need) + 1)) != need)
  {
    need |= need >> 1;
    need |= need >> 2;
    need |= need >> 4;
    for (i = sizeof(need), x = 8; i; i >>= 1, x <<= 1)
      need |= need >> x;

    need += 1;
  }
  for (i = 0, x = need; x; x >>= 1)
    i += 1;

  if (libzahl_pool_n[i])
  {
    libzahl_pool_n[i] -= 1;
    new_idx = libzahl_pool[i][libzahl_pool_n[i]];
    memcpy(new, a->chars, a->alloced * (sizeof(zahl_char_t)));
    zfree(a);
    a->chars = &new[new_idx];
  }
  else
  {
    a->chars = realloc(a->chars, need * (sizeof(zahl_char_t)));
    if (!a->chars)
    {
      if (!errno)
      {
        errno = ENOMEM;
      }
      libzahl_error = errno, longjmp(libzahl_jmp_buf, 1);
    }
  }
  a->alloced = need;
}


----------------------------
typedef uint32_t zahl_char_t
----------------------------
***/


